lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +434,1140 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
+          if (isa<ConstantInt>(CI->getArgOperand(Idx))) {
+            if (cast<ConstantInt>(CI->getArgOperand(Idx))->getZExtValue() == 0)
+              CalleeSavedRegs.insert(ABIRegisters[Idx - 3]);
+          }
+        }
+      }
+    } else {
+      Result.insert({ E.first, FunctionEdgeTypeValue::BrokenReturn });
+    }
+  }
+
+  // Neither a return nor a broken return?
+  for (CallBase *Call : callers(IndirectBranchInfoMarker)) {
+    auto *CI = cast<CallInst>(Call);
+    if (CI->getParent()->getParent() == F) {
+      if (isa<ConstantInt>(CI->getArgOperand(0)))
+        continue;
+
+      auto *V = dyn_cast<ConstantInt>(CI->getArgOperand(1));
+      if (WinFSO.has_value() && V != nullptr && V->getZExtValue() == *WinFSO)
+        Result.insert({ CI, FunctionEdgeTypeValue::IndirectTailCall });
+      else
+        Result.insert({ CI, FunctionEdgeTypeValue::LongJmp });
+    }
+  }
+
+  bool FoundRet, FoundBrokenRet;
+  FoundRet = FoundBrokenRet = false;
+  for (const auto &E : Result) {
+    if (E.second == FunctionEdgeTypeValue::Return)
+      FoundRet = true;
+    if (E.second == FunctionEdgeTypeValue::BrokenReturn)
+      FoundBrokenRet = true;
+  }
+
+  FunctionTypeValue Type;
+  if (FoundRet) {
+    Type = FunctionTypeValue::Regular;
+  } else if (FoundBrokenRet) {
+    Type = FunctionTypeValue::Fake;
+  } else {
+#ifdef NORETURN_IS_FIXED
+    Type = FunctionTypeValue::NoReturn;
+#else
+    Type = FunctionTypeValue::Regular;
+#endif
+  }
+
+  // Finally retrieve the clobbered registers and return them.
+  std::erase_if(ClobberedRegs, [&](const auto &E) {
+    return CalleeSavedRegs.find(E) != CalleeSavedRegs.end();
+  });
+
+  // Finalize CFG for the model.
+  for (const auto &[CI, EdgeValue] : Result) {
+    auto PC = MetaAddress::fromConstant(CI->getArgOperand(2));
+    auto &Block = CFG.at(PC);
+    Block.Successors.insert(makeEdge(MetaAddress::invalid(), EdgeValue));
+  }
+
+  // We say that a register is callee-saved when, besides being preserved by the
+  // callee, there is at least a write onto this register.
+  std::set<GlobalVariable *> ActualCalleeSavedRegs;
+  std::set_intersection(CalleeSavedRegs.begin(),
+                        CalleeSavedRegs.end(),
+                        WrittenRegisters.begin(),
+                        WrittenRegisters.end(),
+                        std::inserter(ActualCalleeSavedRegs,
+                                      ActualCalleeSavedRegs.begin()));
+
+  // Union between effective callee-saved registers and SP.
+  std::set<GlobalVariable *> CalleeSavedAndStackPointerRegs = {
+    ActualCalleeSavedRegs
+  };
+  CalleeSavedAndStackPointerRegs.insert(GCBI->spReg());
+
+  // Refine ABI analyses by suppressing callee-saved and stack pointer registers
+  // from the final results.
+  for (const auto &Reg : CalleeSavedAndStackPointerRegs) {
+    if (ABIResults.Arguments.count(Reg) != 0)
+      ABIResults.Arguments[Reg] = model::RegisterState::Values::No;
```

Comment:
Double lookup.
```suggestion
    auto It = ABIResults.Arguments.find(Reg);
    if (It != ABIResults.Arguments.end())
      It->second = model::RegisterState::Values::No;
```

Also, this must be done for return values too!